x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/autoscaling/AutoscalerService.java

```diff
@@ -0,0 +1,218 @@
+        }
+    }
+
+    private void scheduleNextTrigger() {
+        try {
+            cancellable = threadPool.schedule(this::trigger, TimeValue.timeValueSeconds(TIME_INTERVAL_SECONDS), threadPool.generic());
+        } catch (EsRejectedExecutionException e) {
+            if (e.isExecutorShutdown() == false) {
+                throw e;
+            }
+        }
+    }
+
+    private synchronized void trigger() {
+        scheduleNextTrigger();
+        getDeploymentStats(ActionListener.wrap(this::processDeploymentStats, e -> logger.warn("Error in inference autoscaling", e)));
+    }
+
+    private void getDeploymentStats(ActionListener<GetDeploymentStatsAction.Response> processDeploymentStats) {
+        String deploymentIds = String.join(",", autoscalers.keySet());
+        ClientHelper.executeAsyncWithOrigin(
+            client,
+            ClientHelper.ML_ORIGIN,
+            GetDeploymentStatsAction.INSTANCE,
+            new GetDeploymentStatsAction.Request(deploymentIds),
+            processDeploymentStats
+        );
+    }
+
+    private void processDeploymentStats(GetDeploymentStatsAction.Response statsResponse) {
+        Map<String, Stats> recentStatsByDeployment = new HashMap<>();
+        Map<String, Integer> numberOfAllocations = new HashMap<>();
+
+        for (AssignmentStats assignmentStats : statsResponse.getStats().results()) {
+            numberOfAllocations.put(assignmentStats.getDeploymentId(), assignmentStats.getNumberOfAllocations());
+            for (AssignmentStats.NodeStats nodeStats : assignmentStats.getNodeStats()) {
+                String statsId = assignmentStats.getDeploymentId() + "@" + nodeStats.getNode().getId();
+                Stats lastStats = lastInferenceStatsByDeploymentNode.get(statsId);
+                Stats nextStats = new Stats(
+                    nodeStats.getInferenceCount().orElse(0L),
+                    nodeStats.getPendingCount() == null ? 0 : nodeStats.getPendingCount(),
+                    nodeStats.getErrorCount() + nodeStats.getTimeoutCount() + nodeStats.getRejectedExecutionCount(),
+                    nodeStats.getAvgInferenceTime().orElse(0.0) / 1000.0
+                );
+                lastInferenceStatsByDeploymentNode.put(statsId, nextStats);
+
+                Stats recentStats = (lastStats == null ? nextStats : nextStats.sub(lastStats));
+                recentStatsByDeployment.compute(
+                    assignmentStats.getDeploymentId(),
+                    (key, value) -> value == null ? recentStats : value.add(recentStats)
+                );
+            }
+        }
+
+        for (Map.Entry<String, Stats> deploymentAndStats : recentStatsByDeployment.entrySet()) {
+            String deploymentId = deploymentAndStats.getKey();
+            Stats stats = deploymentAndStats.getValue();
+            Autoscaler autoscaler = autoscalers.get(deploymentId);
+            autoscaler.process(stats, TIME_INTERVAL_SECONDS, numberOfAllocations.get(deploymentId));
```

Comment:
If you take my previous suggestion about scheduling then track the time interval locally between these calls rather than use the fixed `TIME_INTERVAL_SECONDS` . There will be a lot of variance in the time to retrieve stats etc